### ISSUE TITLE
GPU CI tests failing

### ISSUE THREAD

**Issue by ivirshup** [scverse/anndata#1410]
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of anndata.
- [X] (optional) I have confirmed this bug exists on the master branch of anndata.

### Report

Still investigating

* Array api compat upgraded
* Taking a really long time to run the failing tests

### Versions

<details>
<summary> </summary>

```python
Name              Version   Build                Channel    

────────────────────────────────────────────────────────────────

  _libgcc_mutex     0.1       conda_forge          conda-forge

  _openmp_mutex     4.5       2_gnu                conda-forge

  bzip2             1.0.8     hd590300_5           conda-forge

  ca-certificates   2024.2.2  hbcca054_0           conda-forge

  colorama          0.4.6     pyhd8ed1ab_0         conda-forge

  coverage          7.4.3     py312h98912ed_1      conda-forge

  cuda-version      11.8      h70ddcb2_3           conda-forge

  cudatoolkit       11.8.0    h4ba93d1_13          conda-forge

  cupy              13.0.0    py312h8e83189_3      conda-forge

  cupy-core         13.0.0    py312hd7a312d_3      conda-forge

  exceptiongroup    1.2.0     pyhd8ed1ab_2         conda-forge

  execnet           2.0.2     pyhd8ed1ab_0         conda-forge

  fastrlock         0.8.2     py312h30efb56_2      conda-forge

  iniconfig         2.0.0     pyhd8ed1ab_0         conda-forge

  ld_impl_linux-64  2.40      h41732ed_0           conda-forge

  libblas           3.9.0     21_linux64_openblas  conda-forge

  libcblas          3.9.0     21_linux64_openblas  conda-forge

  libexpat          2.6.1     h59595ed_0           conda-forge

  libffi            3.4.2     h7f98852_5           conda-forge

  libgcc-ng         13.2.0    h807b86a_5           conda-forge

  libgfortran-ng    13.2.0    h69a702a_5           conda-forge

  libgfortran5      13.2.0    ha4646dd_5           conda-forge

  libgomp           13.2.0    h807b86a_5           conda-forge

  liblapack         3.9.0     21_linux64_openblas  conda-forge

  libllvm14         14.0.6    hcd5def8_4           conda-forge

  libnsl            2.0.1     hd590300_0           conda-forge

  libopenblas       0.3.26    pthreads_h413a1c8_0  conda-forge

  libsqlite         3.45.1    h2797004_0           conda-forge

  libstdcxx-ng      13.2.0    h7e041cc_5           conda-forge

  libuuid           2.38.1    h0b41bf4_0           conda-forge

  libxcrypt         4.4.36    hd590300_1           conda-forge

  libzlib           1.2.13    hd590300_5           conda-forge

  llvmlite          0.42.0    py312hb06c811_1      conda-forge

  ncurses           6.4       h59595ed_2           conda-forge

  numba             0.59.0    py312hacefee8_1      conda-forge

  numpy             1.26.4    py312heda63a1_0      conda-forge

  openssl           3.2.1     hd590300_0           conda-forge

  packaging         24.0      pyhd8ed1ab_0         conda-forge

  pip               24.0      pyhd8ed1ab_0         conda-forge

  pluggy            1.4.0     pyhd8ed1ab_0         conda-forge

  pytest            8.1.1     pyhd8ed1ab_0         conda-forge

  pytest-cov        4.1.0     pyhd8ed1ab_0         conda-forge

  pytest-xdist      3.5.0     pyhd8ed1ab_0         conda-forge

  python            3.12.2    hab00c5b_0_cpython   conda-forge

  python_abi        3.12      4_cp312              conda-forge

  readline          8.2       h8228510_1           conda-forge

  setuptools        69.1.1    pyhd8ed1ab_0         conda-forge

  tk                8.6.13    noxft_h4845f30_101   conda-forge

  toml              0.10.2    pyhd8ed1ab_0         conda-forge

  tomli             2.0.1     pyhd8ed1ab_0         conda-forge

  tzdata            2024a     h0c530f3_0           conda-forge

  wheel             0.42.0    pyhd8ed1ab_0         conda-forge

  xz                5.2.6     h166bdaf_0           conda-forge

Package            Version

------------------ ---------------------

anndata            0.11.0.dev81+g12878b3

anyio              4.3.0

array_api_compat   1.5

asciitree          0.3.3

awkward            2.6.2

awkward-cpp        30

boltons            23.1.1

certifi            2024.2.2

click              8.1.7

cloudpickle        3.0.0

colorama           0.4.6

contourpy          1.2.0

coverage           7.4.3

cupy               13.0.0

cycler             0.12.1

dask               2024.2.1

distributed        2024.2.1

et-xmlfile         1.1.0

exceptiongroup     1.2.0

execnet            2.0.2

fasteners          0.19

fastrlock          0.8.2

fonttools          4.49.0

fsspec             2024.2.0

h11                0.14.0

h5py               3.10.0

httpcore           1.0.4

httpx              0.27.0

idna               3.6

importlib_metadata 7.0.2

iniconfig          2.0.0

Jinja2             3.1.3

joblib             1.3.2

kiwisolver         1.4.5

linkify-it-py      2.0.3

llvmlite           0.42.0

locket             1.0.0

loompy             3.0.7

markdown-it-py     3.0.0

MarkupSafe         2.1.5

matplotlib         3.8.3

mdit-py-plugins    0.4.0

mdurl              0.1.2

memray             1.11.0

msgpack            1.0.8

natsort            8.4.0

networkx           3.2.1

numba              0.59.0

numcodecs          0.12.1

numpy              1.26.4

numpy-groupies     0.10.2

openpyxl           3.1.2

packaging          24.0

pandas             2.2.1

partd              1.4.1

patsy              0.5.6

pillow             10.2.0

pip                24.0

pluggy             1.4.0

psutil             5.9.8

pyarrow            15.0.1

Pygments           2.17.2

pynndescent        0.5.11

pyparsing          3.1.2

pytest             8.1.1

pytest-cov         4.1.0

pytest-memray      1.5.0

pytest-mock        3.12.0

pytest-xdist       3.5.0

python-dateutil    2.9.0.post0

pytz               2024.1

PyYAML             6.0.1

rich               13.7.1

scanpy             1.9.8

scikit-learn       1.4.1.post1

scipy              1.12.0

seaborn            0.13.2

session_info       1.0.0

setuptools         69.1.1

setuptools-scm     8.0.4

six                1.16.0

sniffio            1.3.1

sortedcontainers   2.4.0

statsmodels        0.14.1

stdlib-list        0.10.0

tblib              3.0.0

textual            0.52.1

threadpoolctl      3.3.0

toml               0.10.2

tomli              2.0.1

toolz              0.12.1

tornado            6.4

tqdm               4.66.2

typing_extensions  4.10.0

tzdata             2024.1

uc-micro-py        1.0.3

umap-learn         0.5.5

urllib3            2.2.1

wheel              0.42.0

zarr               2.17.1

zict               3.0.0

zipp               3.17.0
```

</details>

**Comment by ivirshup** [comment]
It looks like a bugged release of `array-api-compat`.

In the recently released `array-api-compat` 1.5:

```python
import array_api_compat, pandas as pd, cupy as cp

xp = array_api_compat.array_namespace(cp.ones(5))

type(xp.asarray(pd.Index([1,2,3])))
```

in array-api-compat<1.5:

```
cupy.ndarray
```

In array-api-compat==1.5:

```
pandas.core.indexes.base.Index
```


This has already been reported on the array-api-compat repo and declared a bug, so we'll just pin against this version for now.

* https://github.com/data-apis/array-api-compat/issues/106

**Comment by flying-sheep** [comment]
Fixed since at least array-api-compat 1.7.1, I get `cupy.ndarray`
